x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/ResultsIndexUpgradeService.java

```diff
@@ -0,0 +1,523 @@
+            }
+        });
+
+        chainTaskExecutor.execute(ActionListener.wrap(
+            createIndexResponses -> createIndexListener.onResponse(true),
+            createIndexListener::onFailure
+        ));
+    }
+
+    /**
+     * Makes the indices readonly if it's not set as a readonly yet
+     */
+    private void setReadOnlyBlock(Client client, List<String> indices, ActionListener<AcknowledgedResponse> listener) {
+        Settings settings = Settings.builder().put(IndexMetaData.INDEX_READ_ONLY_SETTING.getKey(), true).build();
+        UpdateSettingsRequest request = new UpdateSettingsRequest(indices.toArray(new String[0]));
+        request.settings(settings);
+        executeAsyncWithOrigin(client.threadPool().getThreadContext(),
+            ML_ORIGIN,
+            request,
+            listener,
+            client.admin().indices()::updateSettings);
+    }
+
+    private void removeReadOnlyBlock(Client client, List<String> indices,
+                                     ActionListener<AcknowledgedResponse> listener) {
+        Settings settings = Settings.builder().put(IndexMetaData.INDEX_READ_ONLY_SETTING.getKey(), false).build();
+        UpdateSettingsRequest request = new UpdateSettingsRequest(indices.toArray(new String[0]));
+        request.settings(settings);
+        executeAsyncWithOrigin(client.threadPool().getThreadContext(),
+            ML_ORIGIN,
+            request,
+            listener,
+            client.admin().indices()::updateSettings);
+    }
+
+    private void reindexOldReadIndicesToNewIndices(Client client,
+                                                   Map<String, String> reindexIndices,
+                                                   ResultsIndexUpgradeAction.Request request,
+                                                   ActionListener<Boolean> listener) {
+        TypedChainTaskExecutor<BulkByScrollResponse> chainTaskExecutor =
+            new TypedChainTaskExecutor<>(
+                client.threadPool().executor(executor),
+                (createIndexResponse) -> { // If there are errors in the reindex, we should stop
+                    Tuple<RestStatus, Throwable> status = getStatusAndCause(createIndexResponse);
+                    return status.v1().equals(RestStatus.OK);
+                },
+                (exception -> true)); // Short circuit and call onFailure for any exception
+
+        List<String> newIndices = new ArrayList<>(reindexIndices.size());
+        reindexIndices.forEach((oldIndex, newIndex) -> {
+            ReindexRequest reindexRequest = new ReindexRequest();
+            reindexRequest.setSourceBatchSize(request.getReindexBatchSize());
+            reindexRequest.setSourceIndices(oldIndex);
+            reindexRequest.setDestIndex(newIndex);
+            reindexRequest.setSourceDocTypes(ElasticsearchMappings.DOC_TYPE);
+            reindexRequest.setDestDocType(ElasticsearchMappings.DOC_TYPE);
```

Comment:
Switching to `_doc` here would mean that in 6.7 after calling the endpoint users who first used ML in 6.x had a type of `doc` on their results index (because the endpoint would do nothing) and users who first used ML in 5.x and got migrated would have a type of `_doc`.  So a search for results in 6.7 might be against an index with type `doc` or type `_doc`.  We should not be searching for specific types in 6.x, but I think introducing the inconsistency is risky - we might end up having to solve a tricky problem in a 6.7 patch.